src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -560,16 +560,26 @@ void MacroAssembler::call_VM_helper(Register oop_result, address entry_point, in
 
 // Maybe emit a call via a trampoline.  If the code cache is small
 // trampolines won't be emitted.
-
 address MacroAssembler::trampoline_call(Address entry, CodeBuffer* cbuf) {
   assert(JavaThread::current()->is_Compiler_thread(), "just checking");
   assert(entry.rspec().type() == relocInfo::runtime_call_type
          || entry.rspec().type() == relocInfo::opt_virtual_call_type
          || entry.rspec().type() == relocInfo::static_call_type
          || entry.rspec().type() == relocInfo::virtual_call_type, "wrong reloc type");
+  assert(CodeCache::contains(entry.target()), "target outside CodeCache");
+
+  // Runtime calls are statically bound.
+  // Once they are generated a call site cannot be changed.
+  // Use target_needs_far_branch to check whether a far branch is needed.
+  // For Java calls a callee can be changed at any time as a result of
+  // callee deoptimization or the callee being C1 compiled has become C2 compiled.
+  // If CodeCache size > 128M, such Java calls must have reserved trampolines for cases
+  // when a new callee address out of 128M range.
+  bool need_trampoline = (entry.rspec().type() == relocInfo::runtime_call_type) ?
+      target_needs_far_branch(entry.target()) : far_branches();
```

Comment:
This is all very confusing too. Can't `target_needs_far_branch()` handle all cases? So we don't need a conditional expression here.

Reply:
I'll update `MacroAssembler::trampoline_call` doc when it should be used. All runtime calls into `CodeCache` must use `far_call`. `far_call` is more compact and faster vs `trampoline_call`:
far call:
```
adrp
add
blr
```
trampoline call:
```
bl trampoline
trampoline:
ldr [embedded]
br
embedded:
8 bytes of target address
```

What do you think?

Reply:
To be clear. `relocInfo::runtime_call_type` is used for stubs calls which are inside CodeCache and for VM runtime calls which are outside CodeCache. Depending on call's distance you can use all 3 types of instruction for all cases - even VM runtime method could be near CodeCache so you can use "near" call. So when you are asking what type of call should be generated you can get all 3 answers.

The optimization is that for other calls (inside CodeCache) you need only 2 types of call: far and near. And `MacroAssembler::far_call()` handles it correctly.

Back to `trampoline_call()` function.
1. The header comment is incorrect: "If the code cache is small trampolines won't be emitted."
2. New `assert(CodeCache::contains(entry.target())` is wrong.
3. I would like to know why `trampoline_call()` could be used for `relocInfo::virtual_call_type` and `relocInfo::opt_virtual_call_type` (based on assert). They should be inside CodeCache and you don't need trampoline for it. Right? Unless you generate code outside CodeCache.
4. I don't understand how call to trampoline is handled (lines 603-607) they do not reference `stub` value returned by `emit_trampoline_stub()`. Why not use `far_call()` to call trampoline code?
5. `target_needs_far_branch()` works only inside CodeCache. You need separate `target_needs_trampoline()` for `relocInfo::runtime_call_type`. New method should check distance to target vs CodeCache boundaries like we do for x86: https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/assembler_x86.cpp#L11701

I think the code could be like this:

```
    if (entry.rspec().type() != relocInfo::runtime_call_type ||
         CodeCache::find_blob(entry.target()) != NULL) {
      // Call inside CodeCache
      far_call(entry, cbuf, rscratch1);
      return pc();
    }
    if (target_needs_trampoline(entry)) { // Check distance from CodeCache
      // Generate trampoline to call outside CodeCache
      stub = emit_trampoline_stub();
      // Call to trampoline
      far_call(stub, cbuf, rscratch1);
    } else {
      // Far call outside CodeCache.
      // Near call can't be used because generated code could be posted in any place in CodeCache when installed.
      adrp
      add
      blr
    }
    return pc();
```

Reply:
Hi Vladimir,
Thank you for the detailed comment. 

> The optimization is that for other calls (inside CodeCache) you need only 2 types of call: far and near. And  MacroAssembler::far_call() handles it correctly.

Yes., I've checked the uses of `trampoline_call`. There are no calls outside CodeCache. They must be replaced with `far_call`. I did this and got a crash of the debug build. Debug builds reduce the branch range from 128M to 2M to be able to test trampolines. Debug builds with the 128M branch range work. With the 2M branch range the debug build fails at:
```
ScopeDesc* CompiledMethod::scope_desc_at(address pc) {
   PcDesc* pd = pc_desc_at(pc);
   guarantee(pd != NULL, "scope must be present");
   return new ScopeDesc(this, pd);
}
```
with the stack:
```
V  [libjvm.so+0xa8f138]  CompiledMethod::scope_desc_at(unsigned char*)+0x40
V  [libjvm.so+0x1567290]  compiledVFrame::compiledVFrame(frame const*, RegisterMap const*, JavaThread*, CompiledMethod*)+0xc8
V  [libjvm.so+0x155f1f0]  vframe::new_vframe(frame const*, RegisterMap const*, JavaThread*)+0xd8
V  [libjvm.so+0xaeb574]  Deoptimization::uncommon_trap_inner(JavaThread*, int)+0x1b4
V  [libjvm.so+0xaeca44]  Deoptimization::uncommon_trap(JavaThread*, int, int)+0x24
```

> I would like to know why trampoline_call() could be used for relocInfo::virtual_call_type and relocInfo::opt_virtual_call_type (based on assert). They should be inside CodeCache and you don't need trampoline for it. Right? 

If CodeCashe is bigger than 128M, we need a trampoline for them.
We generate:
```
bl trampoline
trampoline:
```
Initially, we have a call of `resolve_virtual_call_stub`. If the stub finds a compiled method, it patches:
- `bl` if the found method is near.
- `trampoline` if the found method is far.
For the 240M CodeCache, we patch `trampoline` when a C2 compiled method calls a C1 compiled method or vice verse. When a C2 compiled method calls a C2 compiled method or a C1 compiled method calls a C1 compiled method, we have a normal direct call.
If we used a `far_call` there, we would always have an indirect call because a call site would be:
```
adrp
add
blr
```
So the use of `trampoline_call` here is an optimization.

>I don't understand how call to trampoline is handled (lines 603-607) they do not reference stub value returned by emit_trampoline_stub(). Why not use far_call() to call trampoline code?

And they should not. Here it is `CodeBuffer`, calls are not linked yet. Trampoline code is put into the stub code section of `CodeBuffer`. We insert a self-calling instruction. When we move the generated code into `CodeCache` we link the calls by patching them based on `relocInfo` records. I think during linking we try to bypass trampolines if possible.

Reply:
Yes, we try to bypass trampolines during moving code into `CodeCache`:
```
void Relocation::pd_set_call_destination(address x) {
  assert(is_call(), "should be a call here");
  if (NativeCall::is_call_at(addr())) {
    address trampoline = nativeCall_at(addr())->get_trampoline();
    if (trampoline) {
      nativeCall_at(addr())->set_destination_mt_safe(x, /* assert_lock */false);
      return;
    }
  }
  MacroAssembler::pd_patch_instruction(addr(), x);
  assert(pd_call_destination(addr()) == x, "fail in reloc");
}

void NativeCall::set_destination_mt_safe(address dest, bool assert_lock) {
....
  ResourceMark rm;
  int code_size = NativeInstruction::instruction_size;
  address addr_call = addr_at(0);
  bool reachable = Assembler::reachable_from_branch_at(addr_call, dest);
...
  // Patch the constant in the call's trampoline stub.
  address trampoline_stub_addr = get_trampoline();
...
  // Patch the call.
  if (reachable) {
    set_destination(dest);
  } else {
    assert (trampoline_stub_addr != NULL, "we need a trampoline");
    set_destination(trampoline_stub_addr);
  }
...
}
```

Reply:
So my assumption that `trampoline_call()` could be used for call outside CodeCache is wrong (and the method's comment is correct). I looked on few places where runtime call is generated and `trampoline_call()` indeed is used only for CodeCache (at least in places are looked on):
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/aarch64.ad#L3858

My points 1. and 2. are void. 

Thank you for answering my questions 3. and 4.

That also explains your `need_trampoline` code change.

You answered all my questions. Please, do latest merge and updates if needed so I can review final changes.

Reply:
hi, @vnkozlov , @eastig 
> I would like to know why trampoline_call() could be used for relocInfo::virtual_call_type and relocInfo::opt_virtual_call_type (based on assert). They should be inside CodeCache and you don't need trampoline for it. Right?

I take a look at `vmIntrinsics::does_virtual_dispatch`, it looks like there are 2 special intrinsics could be virtual. 
when c2 calls runtime stubs for them, we actually give up this optimization, right?  Is it possible to include those 2 reloc types as well?

Reply:
Hi Xin(@navyxliu),

> I take a look at vmIntrinsics::does_virtual_dispatch, it looks like there are 2 special intrinsics could be virtual.
when c2 calls runtime stubs for them, we actually give up this optimization, right?

What optimization are you referring to?

Reply:
hi, 

For both LibraryCallKit::inline_native_hashcode(), inline_native_clone(), they may generate runtime calls with rrelocInfo::opt_virtual_call_type.  current trampoline_call() opts out this optimization for them, right?